File: backend/src/ops/ingest.ts
import { add_hsg_memory } from "../memory/hsg";
import { q, transaction, vector_store } from "../core/db";
import { rid, now, j } from "../utils";
import { extractText, ExtractionResult } from "./extract";
import { log } from "../core/log";
import { chunk_text } from "../utils/chunking";
import { embedForSector, vectorToBuffer } from "../memory/embed";

const LG = 8000,
    SEC = 1000; // Reduced default section size for better embedding quality

export interface ingestion_cfg {
    force_root?: boolean;
    sec_sz?: number;
    lg_thresh?: number;
}
export interface IngestionResult {
    root_memory_id: string;
    child_count: number;
    total_tokens: number;
    strategy: "single" | "root-child";
    extraction: ExtractionResult["metadata"];
}

const mkRoot = async (
    txt: string,
    ex: ExtractionResult,
    meta?: Record<string, unknown>,
    user_id?: string | null,
    total_sections: number = 1
) => {
    const sum = txt.length > 800 ? txt.slice(0, 800) + "..." : txt;
    const cnt = `[Document: ${ex.metadata.content_type.toUpperCase()}]\n\n${sum}\n\n[Full content split across ${total_sections} sections]`;
    const id = rid(),
        ts = now();

    // Generate embedding for root to make it searchable
    let vec_buf = null;
    let dim = 0;
    try {
        const vec = await embedForSector(cnt, "reflective");
        vec_buf = vectorToBuffer(vec);
        dim = vec.length;
    } catch (e) {
        log.warn("Failed to embed root memory", { error: e });
    }

    await transaction.begin();
    try {
        await q.ins_mem.run(
            id,
            user_id || "anonymous", // user_id
            0, // segment (default 0 for root)
            cnt, // content
            "", // simhash (not computed for root meta-doc)
            "reflective", // primary_sector
            j([]), // tags
            j({
                ...meta,
                ...ex.metadata,
                is_root: true,
                ingestion_strategy: "root-child",
                ingested_at: ts,
            }), // meta
            ts, // created_at
            ts, // updated_at
            ts, // last_seen_at
            1.0, // salience
            0.1, // decay_lambda
            1, // version
            dim || null, // mean_dim
            vec_buf, // mean_vec
            null, // compressed_vec
            0 // feedback_score
        );

        if (vec_buf && dim > 0) {
            await vector_store.storeVector(
                id,
                "reflective",
                Array.from(new Float32Array(vec_buf.buffer, vec_buf.byteOffset, vec_buf.byteLength / 4)),
                dim,
                user_id || "anonymous"
            );
        }

        await transaction.commit();
        return id;
    } catch (e) {
        log.error("Root creation failed", { error: e });
        await transaction.rollback();
        throw e;
    }
};

const mkChild = async (
    txt: string,
    idx: number,
    tot: number,
    rid: string,
    meta?: Record<string, unknown>,
    user_id?: string | null,
) => {
    const r = await add_hsg_memory(
        txt,
        j([]),
        {
            ...meta,
            is_child: true,
            section_index: idx,
            total_sections: tot,
            parent_id: rid,
        },
        user_id || undefined,
    );
    return r.id;
};

const link = async (
    rid: string,
    cid: string,
    idx: number,
    user_id?: string | null,
) => {
    const ts = now();
    await transaction.begin();
    try {
        await q.ins_waypoint.run(rid, cid, user_id || "anonymous", 1.0, ts, ts);
        await transaction.commit();
        log.info(
            `Linked root to child`, { root: rid.slice(0, 8), child: cid.slice(0, 8), section: idx }
        );
    } catch (e) {
        await transaction.rollback();
        log.error(`Link failed for section`, { section: idx, error: e });
        throw e;
    }
};

export async function ingestDocument(
    t: string,
    data: string | Buffer,
    meta?: Record<string, unknown>,
    cfg?: ingestion_cfg,
    user_id?: string | null,
): Promise<IngestionResult> {
    const th = cfg?.lg_thresh || LG,
        sz = cfg?.sec_sz || SEC;
    const ex = await extractText(t, data);
    const { text, metadata: exMeta } = ex;
    const useRC = cfg?.force_root || exMeta.estimated_tokens > th;

    if (!useRC) {
        const r = await add_hsg_memory(
            text,
            j([]),
            {
                ...meta,
                ...exMeta,
                ingestion_strategy: "single",
                ingested_at: now(),
            },
            user_id || undefined,
        );
        return {
            root_memory_id: r.id,
            child_count: 0,
            total_tokens: exMeta.estimated_tokens,
            strategy: "single",
            extraction: exMeta,
        };
    }

    // Use standardized chunking logic
    // sec_sz typically refers to characters in this context based on legacy usage,
    // but chunk_text uses an estimated 'tgt' tokens.
    // Assuming 1 token ~ 4 chars, we convert sec_sz to token target.
    const tgt_tokens = Math.floor(sz / 4);
    const chunks = chunk_text(text, tgt_tokens);
    const secs = chunks.map(c => c.text);

    log.info("Ingesting document", { tokens: exMeta.estimated_tokens, sections: secs.length });

    let rid: string;
    const cids: string[] = [];

    try {
        rid = await mkRoot(text, ex, meta, user_id, secs.length);
        log.info("Root memory created", { id: rid });
        for (let i = 0; i < secs.length; i++) {
            try {
                const cid = await mkChild(
                    secs[i],
                    i,
                    secs.length,
                    rid,
                    meta,
                    user_id,
                );
                cids.push(cid);
                await link(rid, cid, i, user_id);
                log.info(
                    `Section processed`, { index: i + 1, total: secs.length, id: cid }
                );
            } catch (e) {
                log.error(
                    `Section processing failed`, { index: i + 1, total: secs.length, error: e }
                );
                throw e;
            }
        }
        log.info(
            "Ingestion completed", { root: rid, child_count: cids.length }
        );
        return {
            root_memory_id: rid,
            child_count: secs.length,
            total_tokens: exMeta.estimated_tokens,
            strategy: "root-child",
            extraction: exMeta,
        };
    } catch (e) {
        log.error("Document ingestion failed", { error: e });
        throw e;
    }
}

export async function ingestURL(
    url: string,
    meta?: Record<string, unknown>,
    cfg?: ingestion_cfg,
    user_id?: string | null,
): Promise<IngestionResult> {
    const { extractURL } = await import("./extract");
    const ex = await extractURL(url);
    const th = cfg?.lg_thresh || LG,
        sz = cfg?.sec_sz || SEC;
    const useRC = cfg?.force_root || ex.metadata.estimated_tokens > th;

    if (!useRC) {
        const r = await add_hsg_memory(
            ex.text,
            j([]),
            {
                ...meta,
                ...ex.metadata,
                ingestion_strategy: "single",
                ingested_at: now(),
            },
            user_id || undefined,
        );
        return {
            root_memory_id: r.id,
            child_count: 0,
            total_tokens: ex.metadata.estimated_tokens,
            strategy: "single",
            extraction: ex.metadata,
        };
    }

    const tgt_tokens = Math.floor(sz / 4);
    const chunks = chunk_text(ex.text, tgt_tokens);
    const secs = chunks.map(c => c.text);

    log.info("Ingesting URL", { url, tokens: ex.metadata.estimated_tokens, sections: secs.length });

    let rid: string;
    const cids: string[] = [];

    try {
        rid = await mkRoot(ex.text, ex, { ...meta, source_url: url }, user_id, secs.length);
        log.info("Root memory created for URL", { id: rid });
        for (let i = 0; i < secs.length; i++) {
            try {
                const cid = await mkChild(
                    secs[i],
                    i,
                    secs.length,
                    rid,
                    { ...meta, source_url: url },
                    user_id,
                );
                cids.push(cid);
                await link(rid, cid, i, user_id);
                log.info(
                    `URL Section processed`, { index: i + 1, total: secs.length, id: cid }
                );
            } catch (e) {
                log.error(
                    `URL Section processing failed`, { index: i + 1, total: secs.length, error: e }
                );
                throw e;
            }
        }
        log.info(
            "URL Ingestion completed", { root: rid, child_count: cids.length }
        );
        return {
            root_memory_id: rid,
            child_count: secs.length,
            total_tokens: ex.metadata.estimated_tokens,
            strategy: "root-child",
            extraction: ex.metadata,
        };
    } catch (e) {
        log.error("URL ingestion failed", { error: e });
        throw e;
    }
}

File: packages/openmemory-py/src/openmemory/ops/ingest.py
import asyncio
import json
import logging
import uuid
import time
from typing import List, Dict, Any, Optional

from ..core.db import q, db
from ..memory.hsg import add_hsg_memory
from .extract import extract_text, extract_url

from ..utils.chunking import chunk_text

logger = logging.getLogger("openmemory.ops.ingest")

from ..core.security import get_encryption

# Constants for ingestion strategy
LG_THRESH = 8000
SEC_SIZE = 3000


async def mk_root(
    txt: str, ex: Dict, meta: Optional[Dict] = None, user_id: Optional[str] = None, tags: Optional[str] = None
) -> str:
    """
    Create a root memory item that serves as a container for a large document.

    Args:
        txt: Full document text.
        ex: Extraction results containing metadata.
        meta: Extra metadata to attach.
        user_id: Owner of the memory.

    Returns:
        The ID of the created root memory.
    """
    summ = txt[:500] + "..." if len(txt) > 500 else txt
    ctype = ex["metadata"]["content_type"].upper()
    sec_count = int(len(txt) / SEC_SIZE) + 1
    content = f"[Document: {ctype}]\n\n{summ}\n\n[Full content split across {sec_count} sections]"

    mid = str(uuid.uuid4())
    ts = int(time.time()*1000)

    full_meta = (meta or {}).copy()
    full_meta.update(ex["metadata"])

    try:
        await q.ins_mem(
            id=mid,
            user_id=user_id or "anonymous",
            segment=1,
            content=get_encryption().encrypt(content),
            primary_sector="reflective",
            tags=tags or json.dumps([]),
            meta=json.dumps(full_meta, default=str),
            created_at=ts,
            updated_at=ts,
            last_seen_at=ts,
            salience=1.0,
            decay_lambda=0.1,
            version=1,
            feedback_score=0,
            commit=False # Defer to parent transaction
        )
        return mid
    except Exception as e:
        logger.error(f"Failed to create root memory: {e}")
        raise e


async def mk_child(
    txt: str,
    idx: int,
    tot: int,
    p_id: str,
    meta: Optional[Dict] = None,
    user_id: Optional[str] = None,
    tags: Optional[str] = None,
) -> str:
    """
    Create a child memory item representing a section of a document.

    Args:
        txt: Section text.
        idx: Index of this section.
        tot: Total number of sections.
        p_id: Parent root ID.
        meta: Extra metadata.
        user_id: Owner ID.

    Returns:
        The ID of the created child memory.
    """
    m = meta or {}
    m.update({
        "is_child": True,
        "section_index": idx,
        "total_sections": tot,
        "parent_id": p_id
    })
    r = await add_hsg_memory(txt, tags or json.dumps([]), m, user_id, commit=False)
    return r["id"]


async def link(
    root_id_val: str, child_id: str, idx: int, user_id: Optional[str] = None
):
    """Create a waypoint link between a root and its child."""
    ts = int(time.time()*1000)
    t = q.tables
    await db.async_execute(f"INSERT INTO {t['waypoints']}(src_id,dst_id,user_id,weight,created_at,updated_at) VALUES (?,?,?,?,?,?)",
               (root_id_val, child_id, user_id or "anonymous", 1.0, ts, ts))


async def _execute_single_strategy(text: str, tags: str, meta: Dict, user_id: Optional[str]) -> Dict[str, Any]:
    """Internal helper for single memory ingestion."""
    r = await add_hsg_memory(text, tags, meta, user_id)
    return {
        "root_memory_id": r["id"],
        "child_count": 0,
        "strategy": "single",
        "extraction": meta
    }

async def _execute_root_child_strategy(
    text: str, 
    chunks: List[Dict], 
    ex: Dict, 
    meta: Dict, 
    user_id: Optional[str], 
    tags: str
) -> Dict[str, Any]:
    """Internal helper for root-child memory ingestion."""
    async with db.transaction():
        rid_val = await mk_root(text, ex, meta, user_id, tags=tags)
        for i, c in enumerate(chunks):
            cid = await mk_child(c["text"], i, len(chunks), rid_val, meta, user_id, tags=tags)
            await link(rid_val, cid, i, user_id)

        return {
            "root_memory_id": rid_val,
            "child_count": len(chunks),
            "strategy": "root-child",
            "extraction": ex["metadata"]
        }

async def ingest_document(
    t: str,
    data: Any,
    meta: Optional[Dict] = None,
    cfg: Optional[Dict] = None,
    user_id: Optional[str] = None,
    tags: Optional[list] = None,
) -> Dict[str, Any]:
    """
    Ingest a document, automatically deciding between single or root-child strategy.
    
    Args:
        t: Source text or content descriptor.
        data: Raw data (bytes or str).
        meta: Optional metadata.
        cfg: Configuration overrides (lg_thresh, sec_sz, etc).
        user_id: Target user ID.
        tags: Optional tags list.
    """
    th = cfg.get("lg_thresh", LG_THRESH) if cfg else LG_THRESH
    sz = cfg.get("sec_sz", SEC_SIZE) if cfg else SEC_SIZE

    if not user_id and meta and "user_id" in meta:
        user_id = meta["user_id"]

    ex = await extract_text(t, data, user_id=user_id)
    text = ex["text"]
    ex_meta = ex["metadata"]
    est_tok = ex_meta["estimated_tokens"]

    use_rc = (cfg and cfg.get("force_root")) or est_tok > th
    tags_json = json.dumps(tags or [])
    
    m = (meta or {}).copy()
    m.update(ex_meta)
    m.update({"ingested_at": int(time.time()*1000)})

    if not use_rc:
        m["ingestion_strategy"] = "single"
        res = await _execute_single_strategy(text, tags_json, m, user_id)
        res["total_tokens"] = est_tok
        return res

    # Use robust chunking for splitting
    chunks = chunk_text(text, tgt=sz, ovr=cfg.get("overlap", 0.1) if cfg else 0.1)
    logger.info(f"Splitting text into {len(chunks)} sections for ingest")
    
    m["ingestion_strategy"] = "root-child"
    try:
        res = await _execute_root_child_strategy(text, chunks, ex, m, user_id, tags_json)
        res["total_tokens"] = est_tok
        return res
    except Exception as e:
        logger.exception(f"Ingest document failed: {e}")
        raise e


async def ingest_url(
    url: str,
    meta: Optional[Dict] = None,
    cfg: Optional[Dict] = None,
    user_id: Optional[str] = None,
    tags: Optional[list] = None,
) -> Dict[str, Any]:
    """
    Ingest text content from a URL with automatic strategy selection.
    """
    ex = await extract_url(url, user_id=user_id)
    
    th = cfg.get("lg_thresh", LG_THRESH) if cfg else LG_THRESH
    sz = cfg.get("sec_sz", SEC_SIZE) if cfg else SEC_SIZE
    est_tok = ex["metadata"]["estimated_tokens"]

    use_rc = (cfg and cfg.get("force_root")) or est_tok > th
    tags_json = json.dumps(tags or [])
    
    m = (meta or {}).copy()
    m.update(ex["metadata"])
    m.update({"source_url": url, "ingested_at": int(time.time()*1000)})

    if not use_rc:
        m["ingestion_strategy"] = "single"
        res = await _execute_single_strategy(ex["text"], tags_json, m, user_id)
        res["total_tokens"] = est_tok
        return res

    # Use robust chunking
    chunks = chunk_text(ex["text"], tgt=sz, ovr=cfg.get("overlap", 0.1) if cfg else 0.1)
    
    m["ingestion_strategy"] = "root-child"
    try:
        res = await _execute_root_child_strategy(ex["text"], chunks, ex, m, user_id, tags_json)
        res["total_tokens"] = est_tok
        return res
    except Exception as e:
        logger.error(f"URL ingest failed for {url}: {e}", exc_info=True)
        raise e
